File: php/src/VedaDB/Wire/Vbp/VBPConnection.php
<?php
declare(strict_types=1);

namespace VedaDB\Wire\Vbp;

use VedaDB\Wire\Vbp\VBPOpcodes as Ops;

/**
 * High-level synchronous VBP client.
 *
 * Public API:
 *   new VBPConnection(host, port, user, password, db, timeoutSec)
 *     ->connect() returns $this
 *     ->execute(sql, params) returns VBPResult
 *     ->ping() returns array with 'latencyMs' and 'nonce'
 *     ->close()
 *
 * Mirrors the Python + Java POC VBPConnection.
 */
class VBPConnection
{
    public const DEFAULT_VBP_PORT = 6380;

    private string $host;
    private int $port;
    private string $user;
    private string $password;
    private string $database;
    private int $timeoutMs;
    private string $authMechanism;

    private ?VBPMultiplexer $mux = null;
    private int $serverVersion = 0;
    private int $serverCaps = 0;
    private int $nextQueryId = 1;
    private bool $connected = false;

    public function __construct(
        string $host,
        int $port,
        string $user,
        string $password,
        string $database = '',
        int $timeoutSeconds = 30,
        ?string $authMechanism = null,
    ) {
        $this->host = $host;
        $this->port = $port;
        $this->user = $user;
        $this->password = $password;
        $this->database = $database;
        $this->timeoutMs = $timeoutSeconds * 1000;
        $envMech = getenv('VEDADB_VBP_MECH');
        $this->authMechanism = $authMechanism !== null && $authMechanism !== ''
            ? strtoupper($authMechanism)
            : ($envMech !== false && $envMech !== '' ? strtoupper($envMech) : Ops::AUTH_MECH_PLAIN);
    }

    public function connect(): self
    {
        $this->mux = new VBPMultiplexer($this->host, $this->port, $this->timeoutMs);

        // CLIENT_HELLO. The server may reply with [SERVER_READY, AUTH_OK]
        // (dev mode, no auth) or just [SERVER_READY] (auth required).
        $hello = VBPTypeCodec::clientHelloBody(1, 0, $this->user, $this->database, 0, $this->user);
        $reply = $this->mux->call(Ops::OP_CLIENT_HELLO, $hello);
        // Find SERVER_READY in the response frames (dev server may also
        // include a no-auth AUTH_OK in the same response).
        $ready = null;
        foreach ($reply['frames'] as $f) {
            if ($f['op'] === Ops::OP_SERVER_READY) {
                $ready = $f;
                break;
            }
        }
        if ($ready === null) {
            throw new VBPError(
                Ops::SQLSTATE_CONNECTION_FAILURE,
                'expected SERVER_READY in CLIENT_HELLO reply, got '
                . Ops::opcodeName($reply['op']),
            );
        }
        $sr = VBPTypeCodec::parseServerReady($ready['body']);
        $this->serverVersion = $sr['serverVersion'];
        $this->serverCaps = $sr['serverCaps'];

        if ($sr['authRequired']) {
            $this->performAuth();
        }
        $this->connected = true;
        return $this;
    }

    private function performAuth(): void
    {
        assert($this->mux !== null);
        if ($this->authMechanism === Ops::AUTH_MECH_SCRAM_SHA_256) {
            $this->performScramAuth();
            return;
        }
        // PLAIN
        $plain = VBPAuth::plainClientFirst($this->user, $this->password);
        $reply = $this->mux->call(Ops::OP_AUTH_RESPONSE, $plain);
        if ($reply['op'] !== Ops::OP_AUTH_OK) {
            throw new VBPError(
                Ops::SQLSTATE_AUTH_FAILED,
                'auth failed: ' . Ops::opcodeName($reply['op']),
            );
        }
    }

    private function performScramAuth(): void
    {
        assert($this->mux !== null);
        $clientNonce = VBPAuth::generateNonce();
        $clientFirst = VBPAuth::clientFirstMessage($this->user, $clientNonce);
        $reply = $this->mux->call(
            Ops::OP_AUTH_RESPONSE,
            'SCRAM-SHA-256 ' . $clientFirst,
        );
        // Dev server in PLAIN mode may simply return AUTH_OK without a challenge.
        if ($reply['op'] === Ops::OP_AUTH_OK) {
            return;
        }
        if ($reply['op'] !== Ops::OP_AUTH_CHALLENGE) {
            throw new VBPError(
                Ops::SQLSTATE_AUTH_FAILED,
                'expected AUTH_CHALLENGE, got ' . Ops::opcodeName($reply['op']),
            );
        }
        $serverFirst = $reply['body'];
        $state = new VBPScramState($clientNonce);
        $clientFinal = VBPAuth::clientFinalMessage($state, $this->user, $this->password, $serverFirst);
        $reply = $this->mux->call(Ops::OP_AUTH_RESPONSE, $clientFinal);
        if ($reply['op'] !== Ops::OP_AUTH_OK) {
            throw new VBPError(
                Ops::SQLSTATE_AUTH_FAILED,
                'expected AUTH_OK, got ' . Ops::opcodeName($reply['op']),
            );
        }
        // Server-final is not delivered in v1 dev server mode; skip verification.
    }

    /**
     * Execute a query and return a VBPResult.
     *
     * @param ?string[] $params
     */
    public function execute(string $sql, ?array $params = null): VBPResult
    {
        $this->ensureConnected();
        assert($this->mux !== null);
        $qid = $this->nextQueryId++;
        $body = VBPTypeCodec::queryBody($qid, $sql, $params);
        $reply = $this->mux->call(Ops::OP_QUERY, $body);

        $columns = [];
        $colTypes = [];
        $rows = [];
        $tag = '';
        $rowsAffected = 0;
        foreach ($reply['frames'] as $f) {
            switch ($f['op']) {
                case Ops::OP_DATA_CHUNK:
                    $chunk = VBPTypeCodec::parseDataChunk($f['body']);
                    $colTypes = $chunk['colTypes'];
                    // Column names: we don't get them in the body, but the
                    // canonical "SELECT 1 AS n" pattern has a single INT4.
                    // For v1 POC, synthesize "col_0", "col_1", ...
                    if (count($columns) === 0) {
                        for ($i = 0; $i < $chunk['nColumns']; $i++) {
                            $columns[] = 'col_' . $i;
                        }
                    }
                    $rows[] = $chunk['values'];
                    break;
                case Ops::OP_ROWS_FINISHED:
                    $body = $f['body'];
                    if (strlen($body) >= 8) {
                        $u = unpack('VnRows/VnCols', substr($body, 0, 8));
                        $nRows = (int) ($u['nRows'] ?? 0);
                        while (count($rows) < $nRows) {
                            $rows[] = array_fill(0, count($colTypes), null);
                        }
                    }
                    break;
                case Ops::OP_COMMAND_COMPLETE:
                    $body = $f['body'];
                    if (strlen($body) >= 4) {
                        $u = unpack('VtLen', substr($body, 0, 4));
                        $tLen = $u['tLen'] ?? 0;
                        if ($tLen > 0 && strlen($body) >= 4 + $tLen) {
                            $tag = substr($body, 4, $tLen);
                        } else {
                            // Some servers (incl. the dev server) emit a
                            // bare ASCII tag with no length prefix.
                            $tag = $body;
                        }
                        if (strlen($body) >= 4 + $tLen + 8) {
                            $ra = unpack('Prows', substr($body, 4 + $tLen, 8));
                            $rowsAffected = (int) ($ra['rows'] ?? 0);
                        }
                    } else {
                        $tag = $body;
                    }
                    break;
            }
        }
        return new VBPResult($columns, $colTypes, $rows, $tag, (int) $rowsAffected);
    }

    /**
     * Send a PING. Returns latency (ms) and the echoed nonce.
     *
     * @return array{latencyMs:float,nonce:string}
     */
    public function ping(): array
    {
        $this->ensureConnected();
        assert($this->mux !== null);
        $nonce = random_bytes(8);
        $start = microtime(true);
        $reply = $this->mux->call(Ops::OP_PING, $nonce);
        $latency = (microtime(true) - $start) * 1000.0;
        if ($reply['op'] !== Ops::OP_PONG) {
            throw new VBPError(
                Ops::SQLSTATE_FEATURE_NOT_SUPPORTED,
                'expected PONG, got ' . Ops::opcodeName($reply['op']),
            );
        }
        return ['latencyMs' => $latency, 'nonce' => $nonce];
    }

    public function close(): void
    {
        if ($this->mux !== null && $this->connected) {
            try {
                $this->mux->call(Ops::OP_CLOSE, '');
            } catch (\Throwable) {
                // ignore — connection may already be closed
            }
        }
        if ($this->mux !== null) {
            $this->mux->close();
            $this->mux = null;
        }
        $this->connected = false;
    }

    public function __destruct()
    {
        try {
            $this->close();
        } catch (\Throwable) {
            // ignore
        }
    }

    public function getServerVersion(): int
    {
        return $this->serverVersion;
    }

    public function getServerCaps(): int
    {
        return $this->serverCaps;
    }

    public function isConnected(): bool
    {
        return $this->connected;
    }

    private function ensureConnected(): void
    {
        if (!$this->connected || $this->mux === null) {
            throw new VBPException(
                Ops::SQLSTATE_CONNECTION_FAILURE,
                'not connected — call connect() first',
            );
        }
    }
}

File: java/src/main/java/io/vedadb/wire/vbp/VBPConnection.java
package io.vedadb.wire.vbp;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * High-level synchronous VBP client.
 *
 * <p>Constructor: {@code new VBPConnection(host, port, user, password, db, timeoutSeconds)}.
 * <p>Call {@link #connect()} to perform CLIENT_HELLO + AUTH (PLAIN or SCRAM) + AUTH_OK.
 * <p>Then {@link #execute(String, List)} for queries. {@link #ping()} for liveness.
 */
public class VBPConnection implements AutoCloseable {

    public static final int DEFAULT_VBP_PORT = 6380;

    private final String host;
    private final int port;
    private final String user;
    private final String password;
    private final String database;
    private final int timeoutMs;
    private final String authMechanism;

    private VBPMultiplexer mux;
    private int serverVersion = 0;
    private int serverCaps = 0;
    private long sessionTokenLo = 0;
    private long sessionTokenHi = 0;
    private final AtomicInteger nextQueryId = new AtomicInteger(1);

    public VBPConnection(String host, int port, String user, String password,
                         String database, int timeoutSeconds) {
        this(host, port, user, password, database, timeoutSeconds, System.getenv("VEDADB_VBP_MECH"));
    }

    public VBPConnection(String host, int port, String user, String password,
                         String database, int timeoutSeconds, String authMechanism) {
        this.host = host;
        this.port = port;
        this.user = user;
        this.password = password;
        this.database = database == null ? "" : database;
        this.timeoutMs = timeoutSeconds * 1000;
        this.authMechanism = (authMechanism == null || authMechanism.isEmpty())
                ? VBPOpcodes.AUTH_MECH_PLAIN : authMechanism.toUpperCase();
    }

    public VBPConnection connect() {
        try {
            mux = new VBPMultiplexer(host, port, timeoutMs);
        } catch (Exception e) {
            throw new VBPException("08006", "connect failed: " + e.getMessage());
        }
        // CLIENT_HELLO
        byte[] hello = VBPTypeCodec.clientHelloBody(1, 0, user, database, (byte) 0, user);
        VBPFrame ready = mux.call(VBPOpcodes.OP_CLIENT_HELLO, hello);
        if (ready.op != VBPOpcodes.OP_SERVER_READY) {
            throw new VBPError("08006", "expected SERVER_READY, got " + VBPOpcodes.opcodeName(ready.op));
        }
        VBPTypeCodec.ServerReadyParts sr = VBPTypeCodec.parseServerReady(ready.body);
        this.serverVersion = sr.serverVersion;
        this.serverCaps = sr.serverCaps;

        // Auth
        if (sr.authRequired) {
            VBPFrame authResp;
            if (VBPOpcodes.AUTH_MECH_SCRAM_SHA_256.equals(authMechanism)) {
                authResp = performScramAuth(ready, sr);
            } else {
                byte[] plain = VBPAuth.plainClientFirst(user, password);
                authResp = mux.call(VBPOpcodes.OP_AUTH_RESPONSE, plain);
            }
            if (authResp.op != VBPOpcodes.OP_AUTH_OK) {
                throw new VBPError("28000", "auth failed: " + VBPOpcodes.opcodeName(authResp.op));
            }
            VBPTypeCodec.AuthOkParts ok = VBPTypeCodec.parseAuthOk(authResp.body);
            this.sessionTokenLo = ok.sessionTokenLo;
            this.sessionTokenHi = ok.sessionTokenHi;
        }
        return this;
    }

    private VBPFrame performScramAuth(VBPFrame ready, VBPTypeCodec.ServerReadyParts sr) {
        // PLAIN is the default; SCRAM flow is implemented in VBPAuth but the
        // vbp_dev_server uses PLAIN dev-mode auth (auth_required bit is 0).
        // This branch is wired but only used when sr.authRequired is true AND
        // the server actually challenges us.
        byte[] nonceBytes = VBPAuth.generateNonce();
        String clientFirst = VBPAuth.clientFirstMessage(user, nonceBytes);
        VBPFrame challenge = mux.call(VBPOpcodes.OP_AUTH_RESPONSE,
                ("SCRAM-SHA-256 " + clientFirst).getBytes(java.nio.charset.StandardCharsets.UTF_8));
        if (challenge.op == VBPOpcodes.OP_AUTH_OK) return challenge;
        if (challenge.op != VBPOpcodes.OP_AUTH_CHALLENGE) {
            throw new VBPError("28000", "expected AUTH_CHALLENGE, got " + VBPOpcodes.opcodeName(challenge.op));
        }
        String serverFirst = new String(challenge.body, java.nio.charset.StandardCharsets.UTF_8);
        VBPAuth.ScramState state = new VBPAuth.ScramState(nonceBytes);
        String clientFinal = VBPAuth.clientFinalMessage(state, user, password, serverFirst);
        VBPFrame serverFinal = mux.call(VBPOpcodes.OP_AUTH_RESPONSE,
                clientFinal.getBytes(java.nio.charset.StandardCharsets.UTF_8));
        if (serverFinal.op == VBPOpcodes.OP_AUTH_OK) return serverFinal;
        if (serverFinal.op != VBPOpcodes.OP_AUTH_CHALLENGE) {
            throw new VBPError("28000", "expected server-final AUTH_CHALLENGE, got " + VBPOpcodes.opcodeName(serverFinal.op));
        }
        String sf = new String(serverFinal.body, java.nio.charset.StandardCharsets.UTF_8);
        VBPAuth.verifyServerFinal(state, sf);
        // After verify, request a final AUTH_OK.
        VBPFrame ok = mux.call(VBPOpcodes.OP_AUTH_RESPONSE, new byte[0]);
        if (ok.op != VBPOpcodes.OP_AUTH_OK) {
            throw new VBPError("28000", "expected AUTH_OK after SCRAM, got " + VBPOpcodes.opcodeName(ok.op));
        }
        return ok;
    }

    public VBPResult execute(String sql) {
        return execute(sql, Collections.emptyList());
    }

    public VBPResult execute(String sql, List<Object> params) {
        if (mux == null) throw new VBPException("08006", "not connected");
        List<byte[]> envs = new ArrayList<>();
        if (params != null) {
            for (Object p : params) {
                envs.add(encodeParam(p));
            }
        }
        byte[] body = VBPTypeCodec.queryBody(nextQueryId.getAndIncrement(), sql, envs);
        VBPFrame f = mux.call(VBPOpcodes.OP_QUERY, body);
        // f is the first terminal frame. For a SELECT, the multiplexer returned
        // ROWS_FINISHED or COMMAND_COMPLETE (it treats them as terminal). The
        // dev server typically returns DATA_CHUNK then COMMAND_COMPLETE — the
        // multiplexer releases the latch on COMMAND_COMPLETE (the most
        // terminal). For v1 POC, we return a minimal result.
        if (f.op == VBPOpcodes.OP_COMMAND_COMPLETE) {
            return new VBPResult(new ArrayList<>(), new ArrayList<>(),
                    new ArrayList<>(), "OK", 0);
        }
        if (f.op == VBPOpcodes.OP_DATA_CHUNK) {
            VBPTypeCodec.DataChunk dc = VBPTypeCodec.parseDataChunk(f.body);
            List<String> cols = new ArrayList<>();
            List<Integer> colTypes = new ArrayList<>();
            for (int t : dc.colTypes) {
                cols.add(VBPOpcodes.typeName(t));
                colTypes.add(t);
            }
            List<List<Object>> rows = new ArrayList<>();
            rows.add(dc.rowValues);
            return new VBPResult(cols, colTypes, rows, "SELECT", 0);
        }
        return new VBPResult(new ArrayList<>(), new ArrayList<>(),
                new ArrayList<>(), VBPOpcodes.opcodeName(f.op), 0);
    }

    public long ping() {
        if (mux == null) throw new VBPException("08006", "not connected");
        long t0 = System.nanoTime();
        // PING body: u64 nonce (dev server requires ≥8 bytes; spec is silent but
        // the Go reference writes the same nonce back in PONG).
        byte[] nonce = new byte[8];
        java.nio.ByteBuffer.wrap(nonce).order(java.nio.ByteOrder.LITTLE_ENDIAN).putLong(t0);
        VBPFrame pong = mux.call(VBPOpcodes.OP_PING, nonce);
        if (pong.op != VBPOpcodes.OP_PONG) {
            throw new VBPError("08006", "expected PONG, got " + VBPOpcodes.opcodeName(pong.op));
        }
        return (System.nanoTime() - t0) / 1_000_000L;
    }

    private static byte[] encodeParam(Object p) {
        if (p == null) return VBPTypeCodec.inputEnvelopeNull(VBPOpcodes.T_TEXT);
        if (p instanceof Boolean) return VBPTypeCodec.inputEnvelope(VBPOpcodes.T_BOOL, VBPTypeCodec.encodeBool((Boolean) p));
        if (p instanceof Integer) return VBPTypeCodec.inputEnvelope(VBPOpcodes.T_INT4, VBPTypeCodec.encodeInt4((Integer) p));
        if (p instanceof Long) return VBPTypeCodec.inputEnvelope(VBPOpcodes.T_INT8, VBPTypeCodec.encodeInt8((Long) p));
        if (p instanceof Short) return VBPTypeCodec.inputEnvelope(VBPOpcodes.T_INT2, VBPTypeCodec.encodeInt2((Short) p));
        if (p instanceof Float) return VBPTypeCodec.inputEnvelope(VBPOpcodes.T_FLOAT4, VBPTypeCodec.encodeFloat4((Float) p));
        if (p instanceof Double) return VBPTypeCodec.inputEnvelope(VBPOpcodes.T_FLOAT8, VBPTypeCodec.encodeFloat8((Double) p));
        if (p instanceof byte[]) return VBPTypeCodec.inputEnvelope(VBPOpcodes.T_BYTEA, VBPTypeCodec.encodeBytea((byte[]) p));
        return VBPTypeCodec.inputEnvelope(VBPOpcodes.T_TEXT, VBPTypeCodec.encodeText(p.toString()));
    }

    @Override
    public void close() {
        if (mux != null) {
            try {
                mux.call(VBPOpcodes.OP_CLOSE, new byte[0], 0, 2000);
            } catch (Exception ignored) {}
            mux.close();
        }
    }

    public int getServerVersion() { return serverVersion; }
    public int getServerCaps() { return serverCaps; }
    public long getSessionTokenLo() { return sessionTokenLo; }
    public long getSessionTokenHi() { return sessionTokenHi; }
}
